File: src/game/projects/engine.ts
import { addItemToInventoryExact } from '../activity/rewards'
import { applyXp } from '../activity/xp'
import type { GameDatabase } from '../data/types'
import { removeIngredients } from '../production/inventory'
import { applyRaceSkillXp } from '../races/races'
import type { PlayerSave } from '../save/types'
import {
  applyEnchantmentToTarget,
  decodeEnchantTarget,
  eligibleEnchantmentTargets,
} from './enchantments'
import { hasProjectKnowledge } from '../npcs/knowledge'
import { applyBountyProjectProgress } from '../bounties/progress'
import { projectFacilityIdForLookup } from '../production/recipes'
import { applyQuestProcessProgress } from '../quests/progress'
import {
  getEnchantment,
  getProject,
  isCompleteProject,
  isEnchantmentOutput,
  maxProjectQuantity,
  meetsProjectKnowledge,
  meetsProjectSkills,
  projectInputs,
  projectSkillRequirements,
} from './projects'

export type ProjectCompleteResult =
  | {
      ok: true
      save: PlayerSave
      outputLabel: string
      outputQty: number
      xpGained: number
      goldSpent: number
    }
  | { ok: false; reason: string }

export function validateProjectCompletion(
  db: GameDatabase,
  save: PlayerSave,
  projectId: string,
  quantity: number,
  enchantTargetId?: string | null,
): { ok: true } | { ok: false; reason: string } {
  const project = getProject(db, projectId)
  if (!project || !isCompleteProject(project)) {
    return { ok: false, reason: 'That project is not available.' }
  }

  const atLocation = db.Facilities.some(
    (row) =>
      row['Location ID'] === save.currentLocationId &&
      projectFacilityIdForLookup(row['Facility ID']) === project['Facility ID'],
  )
  if (!atLocation) {
    return { ok: false, reason: 'Travel to the required facility first.' }
  }

  if (!meetsProjectKnowledge(db, save, project)) {
    const knowledge = hasProjectKnowledge(db, save, project['Skill ID'])
    if (!knowledge.ok) {
      return {
        ok: false,
        reason: `Speak with the ${knowledge.npcName} to unlock these projects.`,
      }
    }
  }

  if (!meetsProjectSkills(save, project)) {
    const missing = projectSkillRequirements(project).find(
      (requirement) =>
        (save.skills.find((skill) => skill.skillId === requirement.skillId)?.level ?? 1) <
        requirement.level,
    )
    const skillName = missing
      ? (db.Skills.find((skill) => skill['Skill ID'] === missing.skillId)?.['Display Name'] ??
        missing.skillId)
      : 'skill'
    return {
      ok: false,
      reason: missing
        ? `Requires ${skillName} level ${missing.level}.`
        : 'Skill requirements are not met.',
    }
  }

  const crafts = Math.floor(quantity)
  if (crafts <= 0) return { ok: false, reason: 'Choose a quantity of at least 1.' }
  if (crafts > maxProjectQuantity(save, project)) {
    return { ok: false, reason: 'Missing materials or gold for that quantity.' }
  }

  const outputId = project['Output Item / Target ID']
  if (isEnchantmentOutput(outputId)) {
    const enchantment = getEnchantment(db, outputId)
    if (!enchantment || enchantment.Status === 'Needs Data') {
      return { ok: false, reason: 'That enchantment is not ready yet.' }
    }
    if (crafts !== 1) {
      return { ok: false, reason: 'Enchantment projects complete one at a time.' }
    }
    const targets = eligibleEnchantmentTargets(db, save, enchantment)
    if (targets.length === 0) {
      return {
        ok: false,
        reason: 'Select a valid equipped or inventory item to enchant.',
      }
    }
    if (!enchantTargetId || !targets.some((target) => target.id === enchantTargetId)) {
      return { ok: false, reason: 'Choose a valid item to enchant.' }
    }
  } else {
    const item = db.Items.find((row) => row['Item ID'] === outputId)
    if (!item) return { ok: false, reason: 'Project output item is missing from data.' }
  }

  return { ok: true }
}

/** Instantly complete one or more Special Production projects. */
export function completeSpecialProject(
  db: GameDatabase,
  save: PlayerSave,
  projectId: string,
  quantity: number,
  enchantTargetId?: string | null,
  nowMs: number = Date.now(),
): ProjectCompleteResult {
  const validation = validateProjectCompletion(
    db,
    save,
    projectId,
    quantity,
    enchantTargetId,
  )
  if (!validation.ok) return validation

  const project = getProject(db, projectId)!
  const crafts = Math.floor(quantity)
  const inputs = projectInputs(project)
  const withMaterials = removeIngredients(save, inputs, crafts)
  if (!withMaterials) {
    return { ok: false, reason: 'Missing required materials.' }
  }

  const goldCost = project['Gold Cost'] * crafts
  if (withMaterials.gold < goldCost) {
    return { ok: false, reason: 'Not enough gold.' }
  }

  let next: PlayerSave = { ...withMaterials, gold: withMaterials.gold - goldCost }
  const outputId = project['Output Item / Target ID']
  const outputQty = project['Output Quantity'] * crafts
  let outputLabel = project['Display Name']

  if (isEnchantmentOutput(outputId)) {
    const enchantment = getEnchantment(db, outputId)!
    outputLabel = enchantment['Display Name']
    const target = decodeEnchantTarget(enchantTargetId)
    if (!target) return { ok: false, reason: 'Choose a valid item to enchant.' }

    // Inventory indexes can shift when materials are removed; re-resolve by item id.
    let resolved = target
    if (target.kind === 'inventory') {
      const prior = save.inventory[target.index]
      if (!prior || prior.enchantmentId) {
        return { ok: false, reason: 'Choose a valid item to enchant.' }
      }
      const refreshed = next.inventory.findIndex(
        (stack) => stack.itemId === prior.itemId && !stack.enchantmentId,
      )
      if (refreshed < 0) {
        return { ok: false, reason: 'Choose a valid item to enchant.' }
      }
      resolved = { kind: 'inventory', index: refreshed }
    }

    const enchanted = applyEnchantmentToTarget(next, resolved, outputId)
    if (!enchanted) return { ok: false, reason: 'Could not apply the enchantment.' }
    next = enchanted
  } else {
    const granted = addItemToInventoryExact(next, outputId, outputQty)
    if (!granted.ok) return granted
    next = granted.save
    outputLabel =
      db.Items.find((item) => item['Item ID'] === outputId)?.['Display Name'] ?? outputLabel
  }

  const xpTotal = applyRaceSkillXp(db, save, project['Skill ID'], project['XP Reward'] * crafts)
  const xpApplied = applyXp(next, db, project['Skill ID'], xpTotal)
  next = xpApplied.save
  next = applyQuestProcessProgress(db, next, project['Project ID'], crafts)
  next = applyBountyProjectProgress(next, project['Project ID'], crafts, nowMs)

  return {
    ok: true,
    save: next,
    outputLabel,
    outputQty: isEnchantmentOutput(outputId) ? 1 : outputQty,
    xpGained: xpTotal,
    goldSpent: goldCost,
  }
}

File: packages/ik_rules/lib/src/projects/engine.dart
import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../achievements/progress.dart';
import '../activity/xp.dart';
import '../bounties/progress.dart';
import '../inventory/add_items.dart';
import '../js_compat.dart';
import '../npcs/knowledge.dart';
import '../production/inventory.dart';
import '../production/recipes.dart';
import '../quests/progress.dart';
import '../races/races.dart';
import '../save/generated/save_models.dart';
import 'enchantments.dart';
import 'projects.dart';

/// Either a green light or the reason a project cannot be completed.
class ProjectValidation {
  const ProjectValidation.ok() : reason = null;

  const ProjectValidation.failed(this.reason);

  bool get ok => reason == null;
  final String? reason;
}

class ProjectCompleteResult {
  const ProjectCompleteResult.ok({
    required this.save,
    required this.outputLabel,
    required this.outputQty,
    required this.xpGained,
    required this.goldSpent,
  }) : reason = null;

  const ProjectCompleteResult.failed(this.reason)
    : save = null,
      outputLabel = null,
      outputQty = 0,
      xpGained = 0,
      goldSpent = 0;

  bool get ok => reason == null;
  final PlayerSave? save;
  final String? outputLabel;
  final num outputQty;
  final num xpGained;
  final num goldSpent;
  final String? reason;
}

ProjectValidation validateProjectCompletion(
  GameDatabase db,
  PlayerSave save,
  String projectId,
  num quantity, [
  String? enchantTargetId,
]) {
  final project = getProject(db, projectId);
  if (project == null || !isCompleteProject(project)) {
    return const ProjectValidation.failed('That project is not available.');
  }

  final atLocation = db.facilities.any(
    (row) =>
        row.raw['Location ID'] == save.currentLocationId &&
        projectFacilityIdForLookup(jsString(row.raw['Facility ID'])) == project.raw['Facility ID'],
  );
  if (!atLocation) {
    return const ProjectValidation.failed('Travel to the required facility first.');
  }

  if (!meetsProjectKnowledge(db, save, project)) {
    final knowledge = hasProjectKnowledge(db, save, jsString(project.raw['Skill ID']));
    if (!knowledge.ok) {
      return ProjectValidation.failed(
        'Speak with the ${knowledge.npcName} to unlock these projects.',
      );
    }
  }

  if (!meetsProjectSkills(save, project)) {
    final missing = projectSkillRequirements(project).firstWhereOrNull(
      (requirement) => getSkillProgress(save, requirement.skillId).level < requirement.level,
    );
    if (missing == null) {
      return const ProjectValidation.failed('Skill requirements are not met.');
    }
    final displayName = db.skills
        .firstWhereOrNull((skill) => skill.raw['Skill ID'] == missing.skillId)
        ?.raw['Display Name'];
    final skillName = displayName is String ? displayName : missing.skillId;
    return ProjectValidation.failed(
      'Requires $skillName level ${jsNumberToString(missing.level)}.',
    );
  }

  final crafts = quantity.floor();
  if (crafts <= 0) return const ProjectValidation.failed('Choose a quantity of at least 1.');
  if (crafts > maxProjectQuantity(save, project)) {
    return const ProjectValidation.failed('Missing materials or gold for that quantity.');
  }

  final outputId = jsString(project.raw['Output Item / Target ID']);
  if (isEnchantmentOutput(outputId)) {
    final enchantment = getEnchantment(db, outputId);
    if (enchantment == null || enchantment.raw['Status'] == 'Needs Data') {
      return const ProjectValidation.failed('That enchantment is not ready yet.');
    }
    if (crafts != 1) {
      return const ProjectValidation.failed('Enchantment projects complete one at a time.');
    }
    final targets = eligibleEnchantmentTargets(db, save, enchantment);
    if (targets.isEmpty) {
      return const ProjectValidation.failed(
        'Select a valid equipped or inventory item to enchant.',
      );
    }
    if (isBlank(enchantTargetId) || !targets.any((target) => target.id == enchantTargetId)) {
      return const ProjectValidation.failed('Choose a valid item to enchant.');
    }
  } else if (!db.items.any((row) => row.raw['Item ID'] == outputId)) {
    return const ProjectValidation.failed('Project output item is missing from data.');
  }

  return const ProjectValidation.ok();
}

/// Instantly completes one or more Special Production projects.
ProjectCompleteResult completeSpecialProject(
  GameDatabase db,
  PlayerSave save,
  String projectId,
  num quantity, {
  String? enchantTargetId,
  required num nowMs,
}) {
  final validation = validateProjectCompletion(db, save, projectId, quantity, enchantTargetId);
  if (!validation.ok) return ProjectCompleteResult.failed(validation.reason);

  final project = getProject(db, projectId)!;
  final crafts = quantity.floor();
  final withMaterials = removeIngredients(
    save,
    projectInputsForSave(
      save,
      project,
    ).map((input) => RecipeIngredient(itemId: input.itemId, quantity: input.quantity)).toList(),
    crafts,
  );
  if (withMaterials == null) {
    return const ProjectCompleteResult.failed('Missing required materials.');
  }

  final goldCost = jsNumber(project.raw['Gold Cost']) * crafts;
  if (withMaterials.gold < goldCost) {
    return const ProjectCompleteResult.failed('Not enough gold.');
  }

  var next = withMaterials.copyWith(gold: withMaterials.gold - goldCost);
  final outputId = jsString(project.raw['Output Item / Target ID']);
  final outputQty = jsNumber(project.raw['Output Quantity']) * crafts;
  var outputLabel = jsString(project.raw['Display Name']);

  if (isEnchantmentOutput(outputId)) {
    outputLabel = jsString(getEnchantment(db, outputId)!.raw['Display Name']);
    final target = decodeEnchantTarget(enchantTargetId);
    if (target == null) {
      return const ProjectCompleteResult.failed('Choose a valid item to enchant.');
    }

    // Inventory indexes can shift when materials are removed; re-resolve by item id.
    var resolved = target;
    if (target is InventoryEnchantTarget) {
      final prior = target.index >= 0 && target.index < save.inventory.length
          ? save.inventory[target.index]
          : null;
      if (prior == null || isNotBlank(prior.enchantmentId)) {
        return const ProjectCompleteResult.failed('Choose a valid item to enchant.');
      }
      final refreshed = next.inventory.indexWhere(
        (stack) => stack.itemId == prior.itemId && isBlank(stack.enchantmentId),
      );
      if (refreshed < 0) {
        return const ProjectCompleteResult.failed('Choose a valid item to enchant.');
      }
      resolved = InventoryEnchantTarget(refreshed);
    }

    final enchanted = applyEnchantmentToTarget(next, resolved, outputId);
    if (enchanted == null) {
      return const ProjectCompleteResult.failed('Could not apply the enchantment.');
    }
    next = enchanted;
  } else {
    final granted = addItemToInventoryExact(next, outputId, outputQty);
    if (!granted.ok) return ProjectCompleteResult.failed(granted.reason);
    next = granted.save!;
    final displayName = db.items
        .firstWhereOrNull((item) => item.raw['Item ID'] == outputId)
        ?.raw['Display Name'];
    if (displayName is String) outputLabel = displayName;
  }

  final skillId = jsString(project.raw['Skill ID']);
  final xpTotal = applyRaceSkillXp(db, save, skillId, jsNumber(project.raw['XP Reward']) * crafts);
  next = applyXp(next, db, skillId, xpTotal).save;
  next = applyQuestProcessProgress(db, next, jsString(project.raw['Project ID']), crafts);
  next = applyBountyProjectProgress(next, jsString(project.raw['Project ID']), crafts, nowMs);
  next = recordProjectMilestones(db, next, jsString(project.raw['Project ID']), crafts);

  return ProjectCompleteResult.ok(
    save: next,
    outputLabel: outputLabel,
    outputQty: isEnchantmentOutput(outputId) ? 1 : outputQty,
    xpGained: xpTotal,
    goldSpent: goldCost,
  );
}
